validate instrumentation configuration that can break build

sergioboi opened the pull request.

sergioboi committed:
validate instrumentation configuration that can break build

dsvoronin commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationConfiguration.kt:
```diff
@@ -66,6 +66,8 @@ open class InstrumentationConfiguration(val name: String) {
     }
 
     fun data(parentInstrumentationParameters: InstrumentationParameters): Data {
+        require(kubernetesNamespace.isNotBlank())
```
`InstrumentationConfiguration` has special `fun validate()`, can this check go there?

dsvoronin commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationParameters.kt:
```diff
@@ -7,12 +7,18 @@ data class InstrumentationParameters(
-    fun applyParameters(newParameters: Map<String, String>): InstrumentationParameters =
-        InstrumentationParameters(
+    fun applyParameters(newParameters: Map<String, String>): InstrumentationParameters {
+        newParameters.forEach { key, value ->
```
extracted precondition function looks better here

dsvoronin commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationParameters.kt:
```diff
@@ -7,12 +7,18 @@ data class InstrumentationParameters(
-    fun applyParameters(newParameters: Map<String, String>): InstrumentationParameters =
-        InstrumentationParameters(
+    fun applyParameters(newParameters: Map<String, String>): InstrumentationParameters {
+        newParameters.forEach { key, value ->
```
it's not clear to me that instrumentation args are immutable from this point,
i'd expect check as close as possible to runner, but before any reservation job (to prevent resource waste)
but fail fast is also good idea. 
Maybe both? two checks?

sergioboi replied:
I don't want to know about it in many places.
My view is we have a structure InstrumentationParameters and it will guarantee that parameters isn't blank. If we had problems in future with incorrect reusing we will fix this.
Now is constructor parameters validation is missing i will add it

dsvoronin replied:
we modify params once more at least here : https://github.com/avito-tech/avito-android/blob/12b0fdff3962329c8814112a726abd24efc15d3d/subprojects/gradle/runner/service/src/main/kotlin/com/avito/runner/service/worker/device/adb/AdbDevice.kt#L80
and it's already too late to prevent resources reservation

But i got your point, it's ok enough for now

dsvoronin commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/target/TargetConfiguration.kt:
```diff
@@ -40,17 +40,21 @@ open class TargetConfiguration(val name: String) : Serializable {
-            )
-    )
+    fun data(parentInstrumentationParameters: InstrumentationParameters): Data {
+        require(deviceName.isNotBlank())
```
`fun validate()`?

dsvoronin commented on subprojects/gradle/runner/service/src/main/kotlin/com/avito/runner/service/worker/device/adb/AdbDevice.kt:
```diff
@@ -272,8 +272,8 @@ data class AdbDevice(
-                "-w",
-                "-r",
+                "-w", // wait for instrumentation to finish before returning.  Required for test runners.
+                "-r", // raw mode need for parsing
```
👍 for comments

`raw mode (is required/is necessary) for further output parsing`

dsvoronin reviewed (commented):
added some thoughts, take a look pls

dsvoronin committed:
Merge branch 'develop' into mbs-7642

sergioboi committed:
reuse validate function
add InstrumentationParameters.kt initialParams validation
improve AdbDevice.kt comment

YapTeam commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/target/TargetConfiguration.kt:
```diff
@@ -40,25 +40,25 @@ open class TargetConfiguration(val name: String) : Serializable {
-                mapOf("deviceName" to deviceName)
-            )
-    )
+    fun data(parentInstrumentationParameters: InstrumentationParameters): Data {
```
Было же красивше?
